clang-tools-extra/clang-tidy/portability/AvoidPlatformSpecificFundamentalTypesCheck.cpp

```diff
@@ -0,0 +1,254 @@
+  IncludeInserter.registerPreprocessor(PP);
+}
+
+void AvoidPlatformSpecificFundamentalTypesCheck::storeOptions(
+    ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, "WarnOnFloats", WarnOnFloats);
+  Options.store(Opts, "WarnOnInts", WarnOnInts);
+  Options.store(Opts, "WarnOnChars", WarnOnChars);
+  Options.store(Opts, "IncludeStyle", IncludeInserter.getStyle());
+}
+
+static std::string getFloatReplacement(const BuiltinType *BT,
+                                       ASTContext &Context) {
+  const TargetInfo &Target = Context.getTargetInfo();
+
+  auto GetReplacementType = [](unsigned Width) {
+    switch (Width) {
+    // This is ambiguous by default since it could be bfloat16 or float16
+    case 16U:
+      return "";
+    case 32U:
+      return "float32_t";
+    case 64U:
+      return "float64_t";
+    case 128U:
+      return "float128_t";
+    default:
+      return "";
+    }
+  };
+
+  switch (BT->getKind()) {
+  // Not an ambiguous type
+  case BuiltinType::BFloat16:
+    return "bfloat16_t";
+  case BuiltinType::Half:
+    return GetReplacementType(Target.getHalfWidth());
+  case BuiltinType::Float:
+    return GetReplacementType(Target.getFloatWidth());
+  case BuiltinType::Double:
+    return GetReplacementType(Target.getDoubleWidth());
+  default:
+    return "";
+  }
+}
+
+void AvoidPlatformSpecificFundamentalTypesCheck::registerMatchers(
+    MatchFinder *Finder) {
+  auto PlatformSpecificFundamentalType = qualType(
+      allOf(builtinType(),
+            anyOf(WarnOnInts ? isBuiltinInt() : unless(anything()),
+                  WarnOnFloats ? isBuiltinFloat() : unless(anything()),
+                  WarnOnChars ? isBuiltinChar() : unless(anything()))));
+
+  if (!WarnOnInts && !WarnOnFloats && !WarnOnChars)
+    return;
+
+  Finder->addMatcher(
+      varDecl(hasType(PlatformSpecificFundamentalType)).bind("var_decl"), this);
+
+  Finder->addMatcher(
+      functionDecl(returns(PlatformSpecificFundamentalType)).bind("func_decl"),
+      this);
+
+  Finder->addMatcher(
+      parmVarDecl(hasType(PlatformSpecificFundamentalType)).bind("param_decl"),
+      this);
+
+  Finder->addMatcher(
+      fieldDecl(hasType(PlatformSpecificFundamentalType)).bind("field_decl"),
+      this);
+
+  Finder->addMatcher(
+      typedefDecl(hasUnderlyingType(PlatformSpecificFundamentalType))
+          .bind("typedef_decl"),
+      this);
+
+  Finder->addMatcher(typeAliasDecl(hasType(PlatformSpecificFundamentalType))
+                         .bind("alias_decl"),
+                     this);
+}
+
+void AvoidPlatformSpecificFundamentalTypesCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  SourceLocation Loc;
+  QualType QT;
+  SourceRange TypeRange;
+
+  auto SetTypeRange = [&TypeRange](auto Decl) {
+    if (Decl->getTypeSourceInfo())
+      TypeRange = Decl->getTypeSourceInfo()->getTypeLoc().getSourceRange();
+  };
+
+  if (const auto *VD = Result.Nodes.getNodeAs<VarDecl>("var_decl")) {
```

Comment:
Since you only ever look at matches for their location and type, you could name each match the same (`bind`), and do `getNodeAs<ValueDecl>("<common bind name>")`. Then you save all of this repetition.